source/geometries/Next100Vessel.cc

```diff
@@ -357,88 +417,78 @@ namespace nexus {
   }
 
 
-
   G4ThreeVector Next100Vessel::GenerateVertex(const G4String& region) const
   {
     G4ThreeVector vertex(0., 0., 0.);
 
-    // Vertex in the whole VESSEL volume except flanges
+    // Vertex in the whole VESSEL volume
     if (region == "VESSEL") {
       G4double rand = G4UniformRand();
-      if (rand < perc_endcap_vol_) {
-	G4VPhysicalVolume *VertexVolume;
-	do {
-	  vertex = tracking_endcap_gen_->GenerateVertex("VOLUME");  // Tracking endcap
-	  // To check its volume, one needs to rotate and shift the vertex
-	// because the check is done using global coordinates
-	G4ThreeVector glob_vtx(vertex);
-	// First rotate, then shift
-	glob_vtx.rotate(pi, G4ThreeVector(0., 1., 0.));
-	glob_vtx = glob_vtx + G4ThreeVector(0, 0, GetELzCoord());
-	  VertexVolume = geom_navigator_->LocateGlobalPointAndSetup(glob_vtx, 0, false);
-	} while (VertexVolume->GetName() != "VESSEL");
+      if (rand < perc_endcap_vol_) { // Endcaps
+        if (G4UniformRand()<0.5){ // Tracking endcap
+        vertex = tracking_endcap_gen_->GenerateVertex("VOLUME");
+        }
+        else{ // Energy endcap
+          vertex = energy_endcap_gen_->GenerateVertex("VOLUME");
+        }
+      }
+      else if (rand < (perc_endcap_vol_ + perc_ep_flange_vol_)){// Energy flange
+        vertex = energy_flange_gen_->GenerateVertex("WHOLE_VOL");
       }
-      else if (rand > 1. - perc_endcap_vol_) {
-	G4VPhysicalVolume *VertexVolume;
-	do {
-	  vertex = energy_endcap_gen_->GenerateVertex("VOLUME");  // Energy endcap
-	  // To check its volume, one needs to rotate and shift the vertex
-	  // because the check is done using global coordinates
-	  G4ThreeVector glob_vtx(vertex);
-	  // First rotate, then shift
-	  glob_vtx.rotate(pi, G4ThreeVector(0., 1., 0.));
-	  glob_vtx = glob_vtx + G4ThreeVector(0, 0, GetELzCoord());
-	  VertexVolume = geom_navigator_->LocateGlobalPointAndSetup(glob_vtx, 0, false);
-	} while (VertexVolume->GetName() != "VESSEL");
+      else if (rand < (perc_endcap_vol_ + perc_ep_flange_vol_ + perc_tp_flange_vol_)){// Tracking flange
+        vertex = tracking_flange_gen_->GenerateVertex("WHOLE_VOL");
       }
-      else
-	vertex = body_gen_->GenerateVertex("BODY_VOL");  // Body
+      else // Body
+        	vertex = body_gen_->GenerateVertex("WHOLE_VOL");
     }
 
-    // Vertex in FLANGES
-    else if (region == "VESSEL_FLANGES") {
-      if (G4UniformRand() < 0.5)
-      	vertex = tracking_flange_gen_->GenerateVertex("BODY_VOL");
-      else
-      	vertex = energy_flange_gen_->GenerateVertex("BODY_VOL");
+    else if (region == "PORT_1a"){
+      vertex = port_gen_->GenerateVertex("WHOLE_VOL");
+
+      vertex = vertex.rotateX( 90. * deg);
```

Comment:
No, I didnt find any logical way since `G4PVPlacement` and `G4UnionSolid` rotations and `G4ThreeVector` rotations work differently. In the first two cases, when performing a multiple rotation (ie. one after the other) you do it with reference to the rotated axis. Nevertheless, the `G4ThreeVector` references the original axis.